### PR TITLE
[wip]: add usb gadget management to KOReader

### PR DESCRIPTION
**Pull request by pazos**
Related to #2597, but with the idea to support more devices (Kobos/Bq) and other usb gadget modules (like usbnet).

The idea is that every supported device implement its own methods to start/stop mass storage & usbnet. Those methods are called from UsbGadgetPlugIn/UsbGadgetPlugOut, which should be hooked to the appropiate "fake" usb events sent from base.

I wrote all this on the emulator, which AFAIK doesn't generate these events. So I added a couple of entries to simulate usb plug in / usb plug out and see the effects depending on current selected mode.

TODO:
If usbGadgetPlugIn and mode == mass storage we need to disable wireless, buttons, plugins, etc. Almost like before suspending but without actually suspending the device. The device needs to stay waiting for a wake event, which should be triggered when the usb connection with a computer is removed.

I need a bit of help with this.

### CODE REVIEW

**Comment by NiLuJe** [comment]
It's unfortunately somewhat [non-trivial](http://trac.ak-team.com/trac/browser/niluje/Configs/trunk/Kindle/Hacks/USBNetwork/src/usbnet/bin/usbnetwork) and to some extend [device and/or FW specific](http://trac.ak-team.com/trac/browser/niluje/Configs/trunk/Kindle/Touch_Hacks/USBNetwork/src/usbnet/bin/usbnetwork), even on [Kobo](http://trac.ak-team.com/trac/browser/niluje/Configs/trunk/Kindle/Kobo_Hacks/KoboStuff/src/usr/local/stuff/bin/usbnet-toggle.sh) ;).

TL;DR: I really really don't want to maintain that mess in two different places ;).

**Comment by NiLuJe** [comment]
That was just a friendly reminder that it might be more complex that one would hope, but feel free to go ahead ;).

**Comment by NiLuJe** [comment]
The fact that this sort of stuff doesn't panic or hang the kernel is also sometimes a bit of the miracle, like the WiFi handling on Kobo ;).

I still don't know what the hell's going on with the KOA2, for instance, where getting it to switch without crashing is *extremely* finicky... ;).

**Comment by NiLuJe** [comment]
And on Kobo, doing it without those *exact* sleep timing would sometime hang, too.

**Comment by NiLuJe** [comment]
Ah, thanks for that ntx comment, you made my day ;).

**Comment by pazos** [comment]
Did a few tests today on my Cervantes. Usbnet works without problems. Mass storage works as long as we don't have any open documents (ie: works from filemanager)

@Frenzie @poire-z: is there an easy way to check if we're in the reader app and return to filemanager?.

**Comment by Frenzie** [comment]
Generally you would pass `ui` when initiating a module, and then in that module you can test something like `self.ui.document`.

I imagine one of the plugins (most likely statistics) must have code like that already.

**Comment by poire-z** [comment]
As @Frenzie said, eg:
https://github.com/koreader/koreader/blob/26aeb6121c977afec17fce45eaf41aeae567a26f/frontend/apps/reader/modules/readergesture.lua#L20
(or grep -i docless frontend and plugins for other contexts - isDocless = is used from FileManager)

**Comment by poire-z** [comment]
>  Mass storage works as long as we don't have any open documents

What happens for example with crash.log, that is opened and koreader has a handle on it, when /mnt/onboard is exported to usb - then stuff written to it by koreader - then unexported - then written to again: is everything that was written before/during/after there ? What if you write or truncate or remove that crash.log from your computer while usb is mounted?
And if you remove a book over usb, is it really not there when usb is disabled?

In the few occasions I enabled usb (via shell, mostly by error) while koreader was running (forgot to close it), I always ended up with stuff done on one side or the other that has disappeared when I disabled usb. Ended up rebooting and having to check what was there or no more there.

Do you need to umount /mnt/onboard ? Or does the usb gadget stuff works over an already mounted stuff? What's the principle of this usb gadget module and ... why is it called _gadget_ ? :)

**Comment by NiLuJe** [comment]
gadget is the Linux kernel nomenclature for all of the weird USB thingies like that, for... reasons I guess? :D.

**Comment by pazos** [comment]
Great questions @poire-z 

> What happens for example with crash.log, that is opened and koreader has a handle on it, when /mnt/onboard is exported to usb - then stuff written to it by koreader - then unexported - then written to again: is everything that was written before/during/after there ? What if you write or truncate or remove that crash.log from your computer while usb is mounted?

I need to test on my Kobo Aura One (on Cervantes this does not apply because KOReader is running from other partition). Will test and give a proper anwser.

> Do you need to umount /mnt/onboard?

Not really but it is recommended to avoid fs corruption. So, in short, yes, we need to unmount /mnt/onboard. This means we need to close all fds.

> What's the principle of this usb gadget module and ... why is it called gadget ? :)

Most embedded devices came without USB Host capabilities. They use some kind of USB gadget module to communicate to another USB Host. Gadget modules are used for mass storage, network connection or even serial over usb). See https://www.kernel.org/doc/html/v4.16/driver-api/usb/gadget.html

**Comment by NiLuJe** [comment]
If all is well, if we still hold an fd to crash.log, the unmount will fail, and all hell breaks loose ;).

**Comment by poire-z** [comment]
And what about the luajit binary and all the libraries in koreader/libs/ ? Are the "reference" to them different from a file handle? I guess they would prevent umount too.
But if it works for plato, I guess there's no reason for it to not work with koreader...

I remember we talked about it before, and I still feel as https://github.com/koreader/koreader/issues/2597#issuecomment-284216250  (I don't mind if we add usb mass storage and its risky - i would just avoid using it - but I would mind having to install koreader elsewhere :)

**Comment by NiLuJe** [comment]
Which is where a lazy unmount comes in, I think, and you better pray nothing ever actually touch the FS ;).

**Comment by pazos** [comment]
Added a work in progress implementation for BQ Cervantes.

**Comment by pazos** [comment]
> And what about the luajit binary and all the libraries in koreader/libs/ ? Are the "reference" to them different from a file handle? I guess they would prevent umount too.

They were buffered, they won't be read from disk again until restarting.

> f.g., I have no idea what happens if you were to try to update KOReader itself, or touch some of its files during that time (yes, I'm sneaky like that, but, if it's doable, assume someone will try to do it one day, no matter how stupid it sounds).

Most of the time you won't notice the changes until restarting (try to edit some Uimanager stuff from KOReader text editor and see what happens) 

> If all is well, if we still hold an open fd to crash.log (or, well, to anything on the partition to be exported), the unmount will fail, and all hell breaks loose ;).

The crash.log redirection is my main worry, but plato seems to write to dosfsck.log even before mounting /mnt/storage, so it should work.

Other fd are not a problem (If we close the current document) because we drop caches before unmounting.

**Comment by baskerville** [comment]
When the partition is unmounted lazily, writing to an existing file will often fail: the file appears to be modified, but when the partition is remounted, the file will be back to its original state.

My `usb-enable.sh` script first tries to unmount without the `-l` flag. If *Plato* is installed in `/mnt/onboard`, this will always fail.

That's why I've installed *Plato* in `/usr/local`. And since I want to be able to edit the configuration file while the storage is shared, `Settings.toml` is a symbolic link to a file in `/mnt/onboard`.

**Comment by pazos** [comment]
I am not interested anymore in implementing this for kobos (I just sold my KA1 and was fine before mounting in Nickel/plato).

For cervantes I'm using `safemode` which is already a prerequisite for install KOReader. I was thinking about using UsbPlugIn to trigger an exit with code 86 (since 85 is used to restart KOReader), so plug with a computer will show a question popup and, after confirming, KOReader will exit, usb mass storage will be enabled until the home button is pressed. Once the button is pressed KOReader will start again.

**Comment by NiLuJe** [comment]
Yeah, an out-of-the-app solution is probably the safest way to handle this anyway ;).

**Comment by pazos** [comment]
> Yeah, an out-of-the-app solution is probably the safest way to handle this anyway ;).

But I'm not sure how to implement that on KOReader. I see two options:

1. call the binary from bash and release the binary elsewhere. Then ko needs to know if the binary is available before showing the popup.
2. integrate the binary in KOReader. It relies on FBInk.
3. integrate the binary in FBInk, @NiLuJe needs to approve.

I'm an open for all the three options, but see 3 as top notch!

**Comment by NiLuJe** [comment]
I'd say 2., link everything dynamically (that include the FBInk CLI), and ship the FBInk shared lib ;).

**Comment by pazos** [comment]
Closing in favor of #4588, I went for option 1 because is less work for me :p
